File: src/csharp/Grpc.Core/DefaultCallInvoker.cs
#region Copyright notice and license

// Copyright 2015-2016 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System.Threading.Tasks;
using Grpc.Core.Internal;
using Grpc.Core.Utils;

namespace Grpc.Core
{
    /// <summary>
    /// Invokes client RPCs using <see cref="Calls"/>.
    /// </summary>
    public class DefaultCallInvoker : CallInvoker
    {
        readonly Channel channel;

        /// <summary>
        /// Initializes a new instance of the <see cref="Grpc.Core.DefaultCallInvoker"/> class.
        /// </summary>
        /// <param name="channel">Channel to use.</param>
        public DefaultCallInvoker(Channel channel)
        {
            this.channel = GrpcPreconditions.CheckNotNull(channel);
        }

        /// <summary>
        /// Invokes a simple remote call in a blocking fashion.
        /// </summary>
        public override TResponse BlockingUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
        {
            var call = CreateCall(method, host, options);
            return Calls.BlockingUnaryCall(call, request);
        }

        /// <summary>
        /// Invokes a simple remote call asynchronously.
        /// </summary>
        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
        {
            var call = CreateCall(method, host, options);
            return Calls.AsyncUnaryCall(call, request);
        }

        /// <summary>
        /// Invokes a server streaming call asynchronously.
        /// In server streaming scenario, client sends on request and server responds with a stream of responses.
        /// </summary>
        public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options, TRequest request)
        {
            var call = CreateCall(method, host, options);
            return Calls.AsyncServerStreamingCall(call, request);
        }

        /// <summary>
        /// Invokes a client streaming call asynchronously.
        /// In client streaming scenario, client sends a stream of requests and server responds with a single response.
        /// </summary>
        public override AsyncClientStreamingCall<TRequest, TResponse> AsyncClientStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options)
        {
            var call = CreateCall(method, host, options);
            return Calls.AsyncClientStreamingCall(call);
        }

        /// <summary>
        /// Invokes a duplex streaming call asynchronously.
        /// In duplex streaming scenario, client sends a stream of requests and server responds with a stream of responses.
        /// The response stream is completely independent and both side can be sending messages at the same time.
        /// </summary>
        public override AsyncDuplexStreamingCall<TRequest, TResponse> AsyncDuplexStreamingCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options)
        {
            var call = CreateCall(method, host, options);
            return Calls.AsyncDuplexStreamingCall(call);
        }

        /// <summary>Creates call invocation details for given method.</summary>
        protected virtual CallInvocationDetails<TRequest, TResponse> CreateCall<TRequest, TResponse>(Method<TRequest, TResponse> method, string host, CallOptions options)
                where TRequest : class
                where TResponse : class
        {
            return new CallInvocationDetails<TRequest, TResponse>(channel, method, host, options);
        }
    }
}

File: src/php/lib/Grpc/DefaultCallInvoker.php
<?php
/*
 *
 * Copyright 2018 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
namespace Grpc;

use Grpc\Internal\InterceptorChannel;

/**
 * Default call invoker in the gRPC stub.
 */
class DefaultCallInvoker implements CallInvoker
{
    /**
     * @param string $hostname
     * @param array<string,mixed> $opts
     *
     * @return Channel
     */
    public function createChannelFactory($hostname, $opts) {
        return new Channel($hostname, $opts);
    }

    /**
     * @param Channel|InterceptorChannel   $channel
     * @param string $method
     * @param array{ 0: class-string, 1: string}&callable $deserialize A function that deserializes the response
     * @param array<string,mixed> $options
     *
     * @return UnaryCall
     */
    public function UnaryCall($channel, $method, $deserialize, $options) {
        return new UnaryCall($channel, $method, $deserialize, $options);
    }

    /**
     * @param Channel|InterceptorChannel   $channel
     * @param string $method
     * @param array{ 0: class-string, 1: string}&callable $deserialize A function that deserializes the response
     * @param array<string,mixed> $options
     *
     * @return ClientStreamingCall
     */
    public function ClientStreamingCall($channel, $method, $deserialize, $options) {
        return new ClientStreamingCall($channel, $method, $deserialize, $options);
    }

    /**
     * @param Channel|InterceptorChannel   $channel
     * @param string $method
     * @param array{ 0: class-string, 1: string}&callable $deserialize A function that deserializes the response
     * @param array<string,mixed> $options
     *
     * @return ServerStreamingCall
     */
    public function ServerStreamingCall($channel, $method, $deserialize, $options) {
        return new ServerStreamingCall($channel, $method, $deserialize, $options);
    }

    /**
     * @param Channel|InterceptorChannel   $channel
     * @param string $method
     * @param array{ 0: class-string, 1: string}&callable $deserialize A function that deserializes the response
     * @param array<string,mixed> $options
     *
     * @return BidiStreamingCall
     */
    public function BidiStreamingCall($channel, $method, $deserialize, $options) {
        return new BidiStreamingCall($channel, $method, $deserialize, $options);
    }
}

